circuits/lib/query/query.circom

```diff
@@ -21,47 +28,67 @@ template Query (valueArraySize) {
+    signal lt <== LessThan254()([in, value[0]]);
 
-    component gt = GreaterThan254();
-    gt.in[0] <== in;
-    gt.in[1] <== value[0];
+    // GreaterThan
+    signal gt <== GreaterThan254()([in, value[0]]);
 
     // in
-    component inComp = IN(valueArraySize);
-    inComp.in <== in;
-    for(var i = 0; i<valueArraySize; i++){inComp.value[i] <== value[i];}
+    signal inComp <== IN(valueArraySize)(in, value);
 
-    // mux
-    component mux = Mux3();
-    component n2b = Num2Bits(3);
-    n2b.in <== operator;
+    // lte
+    signal lte <== 1 - gt; // lte === !gt
 
-    mux.s[0] <== n2b.out[0];
-    mux.s[1] <== n2b.out[1];
-    mux.s[2] <== n2b.out[2];
+    // gte
+    signal gte <== 1 - lt; // gte === !lt
 
-    mux.c[0] <== 1; // noop, skip execution
-    mux.c[1] <== eq.out;
-    mux.c[2] <== lt.out;
-    mux.c[3] <== gt.out;
-    mux.c[4] <== inComp.out; // in
+    // between (value[0] <= in <= value[1])
+    signal gt2 <== GreaterThan254()([in, value[1]]);
+    signal lte2 <== 1 - gt2; // lte === !gt
+    signal between <== AND()(gte, lte2);
 
-    mux.c[5] <== 1-inComp.out; // nin
+    signal opBits[5] <== Num2Bits(5)(operator); // values 0-15 are query operators, 16-31 - modifiers/computations
 
-    mux.c[6] <== 1-eq.out; // neq
-    mux.c[7] <== 0; // not in use
+    // query operator mux
+    component queryOpSatisfied = Mux4();
+    queryOpSatisfied.s <== [opBits[0], opBits[1], opBits[2], opBits[3]];
+    // We don't use 5th bit (opBits[4]) here; which specifies whether operator is query or
+    // modifier/computation operator. It's used in final mux.
 
-    // output
-    out <== mux.out;
+    queryOpSatisfied.c[0] <== 1; // noop; skip execution
+    queryOpSatisfied.c[1] <== eq;
+    queryOpSatisfied.c[2] <== lt;
+    queryOpSatisfied.c[3] <== gt;
+    queryOpSatisfied.c[4] <== inComp; // in
+    queryOpSatisfied.c[5] <== 1-inComp; // nin
+    queryOpSatisfied.c[6] <== 1-eq; // neq
+    queryOpSatisfied.c[7] <== lte; // lte === !gt
+    queryOpSatisfied.c[8] <== gte; // gte === !lt
+    queryOpSatisfied.c[9] <== between; // between
+    queryOpSatisfied.c[10] <== 0; // not used
+    queryOpSatisfied.c[11] <== 0; // not used
+    queryOpSatisfied.c[12] <== 0; // not used
+    queryOpSatisfied.c[13] <== 0; // not used
+    queryOpSatisfied.c[14] <== 0; // not used
+    queryOpSatisfied.c[15] <== 0; // not used
+
+    // modifier operation validation mux
+    // it only validates that operator number is valid
+    component modifierOpValid = Mux4();
+    modifierOpValid.s <== [opBits[0], opBits[1], opBits[2], opBits[3]];
+    // valid operator: 16 - selective disclosure (16-16 = index 0)
+    // valid operator: 17 - nullify (17-16 = index 1)
+    modifierOpValid.c <== [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+
+    // final output mux
+    out <== Mux1()(
+        s <== opBits[4], // specifies whether operator is query or modifier/computation operator
+        c <== [queryOpSatisfied.out, modifierOpValid.out]
```

Comment:
GreaterThan254 generates a lot of unneeded constraints. It could be `greaterThan(5)([operator, 15])`, but `opBits[4]` already has needed flag. It would make sense to check `> 15` if we would have more than 5 bit values for operator.